### PR TITLE
Round Robin & Curriculum learning

### PR DESCRIPTION
**Pull request by sabraaap**
Both methods (round robin batch and curriculum learning) require us to make a list of datasets for all levels and passing this into the appropriate collator as the dataset (see `train_agent_babyai.py`). Both the `RoundRobinCollator `and `CurriculumCollator` classes then construct collators for the individual datasets and sample from the respective method.

**For round-robin learning:**
-Levels are randomly shuffled when the collator is instantiated, after that the order remains the same
-During each epoch, cycle through the level-specific collators and sample a full batch for the respective level 

**For curriculum learning:**
-Levels are in the order specified in the original paper or ordered according to a list of indices in a custom order, e.g. corresponding to the results from the single-task experiments (which has to be passed on instantiation), and/or in reversed order for the anti-curriculum setting
-We use a weighted random sampler and a set of weights corresponding to the levels to sample mixed batches with sub-episodes from multiple tasks which is based on the number of levels to be included in the batch (
`weight_i = level_i / (n_levels_to_include * (n_levels_to_include + 1) / 2)`
-We introduce a new level per epoch (during the first epoch, batches are made up of only sub-episodes from the first level), until all levels have been introduced and then train again for the same number of epochs - meaning the `n_epochs = n_levels * 2 `(this overrides the `n_epochs` in the args, and this is done in the training script) - _this definitely needs tested, but it seems to be the point where we reach roughly equal n_samples_processed for all levels_
_-Note: I noticed that in the original Collator logic, batches are sampled with the size corresponding to the `"num_episodes"` rather than the "batch_size" arg... not sure if this was intentional or not, especially since it seems to propagate through from the Trainer calls to the collator instance (where `features` is passed to the collator, and we subsequently use the length of `features` as the batch size - however, it turns out that` len(features)` is equal to `args["num_episodes"]` 🤔. I've implemented a workaround for this that ignores `features` but maybe worth looking into and verifying!_

**Evaluation:**
-For both MT settings, we evaluate on all levels (and for sample efficiency, based on the total number of samples processed, rather than only samples processed for the respective task)

_**A few other changes to the code in general in this PR to highlight:**
-The level PutNext is not really a maze (just a wider room) but constructed by using two columns (which in the ood/iid seed logic makes them a maze, and an ood one at that). I've updated the seed logic to account for this and treat PutNext tasks like a single room.
-In preparation of the experiments we will be running, I have added two extra args - `"use_mission" `(set to `False` for one of the feedback ablation experiments and the vanilla DT) and `"use_pretrained"` (set to True for the experiment using the pre-trained GPT-2 model weights - this does not currently have a corresponding implementation in the code yet though)
-Added +1 to the episode end index (as we got an error here when the episode was just one time step long, resulting in start and end being the same)
-Got rid of the `"random_used_action_space_only"` option for random datasets (mostly because this causes issues for the MT case where the used action space differs between tasks)
-Changed dataset generation to assume that SAR(F) belong to the same time step (rather than R(F)SA)_

### CODE REVIEW

File: scripts/train_agent_babyai.py [hunk 1]
```diff
@@ -1,52 +1,88 @@
...
+        log("using cpu")
 else:
-    device = torch.device("cpu")
-    log("using cpu")
+    device = torch.device("mps")
+    log("using mps")
 
-log("creating dataset...")
-dataset = CustomDataset.get_dataset(args)
+if not "single" in args["train_mode"]:
```
Comment (maxtaylordavies) [inline scripts/train_agent_babyai.py:59]: would be great to document somewhere what values are allowed for `train_mode` (apologies if you already did and i missed it)

File: src/collator/collator.py [hunk 2]
```diff
@@ -1,21 +1,23 @@
...
 
 @dataclass
 class Collator:
     def __init__(
         self,
         custom_dataset: CustomDataset,
-        feedback=True,
-        mission=True,
-        context_length=64,
+        args,
```
Comment (maxtaylordavies) [inline src/collator/collator.py:20]: nice refactor, thanks

File: src/collator/collator.py [hunk 3]
```diff
@@ -65,6 +71,9 @@ def __init__(
     def reset_counter(self):
         self.samples_processed = 0
 
+    def update_epoch(self):
+        pass
```
Comment (maxtaylordavies) [inline src/collator/collator.py:75]: is this supposed to just be pass?

File: src/collator/collator.py [hunk 4]
```diff
@@ -153,10 +169,147 @@ def _sample_batch(self, batch_size, random_start=True, full=False, train=True):
...
+        self.samples_processed_by_level = {key: 0 for key in range(len(self.datasets))}
+
+    def reset_weights(self):
+        self.weights = np.zeros(len(self.collators))
+        self.weights[0] = 1
+
+    def update_epoch(self):
+        self._update_weights()
+
+    def _update_weights(self):
```
Comment (maxtaylordavies) [inline src/collator/collator.py:272]: this is cool

File: src/collator/collator.py [hunk 5]
```diff
@@ -153,10 +169,147 @@ def _sample_batch(self, batch_size, random_start=True, full=False, train=True):
...
+            ] += self._count_samples_processed(batch, n_samples)
+
+        if train:
+            self.samples_processed += self._count_samples_processed(mixed_batch)
+        log(self.samples_processed_by_level)
+
+        return mixed_batch
+
+    def __call__(self, features):
+        batch_size = self.args["batch_size"]
```
Comment (maxtaylordavies) [inline src/collator/collator.py:314]: thanks for sorting this

File: src/dataset/custom_dataset.py [hunk 6]
```diff
@@ -404,8 +450,9 @@ def _generate_new_dataset(self):
                         current_episode += 1
                         current_conf_episode += 1
 
-        self.env.close()
-        self._clear_buffer(obs.shape)
+        if hasattr(self, "env"):
+            self.env.close()
+            self._clear_buffer(obs.shape)
```
Comment (maxtaylordavies) [inline src/dataset/custom_dataset.py:455]: should `self._clear_buffer` be indented under `hasattr(self, "env")` like this?

**Review by maxtaylordavies — APPROVED** [review]
Looks good to me, nice job! left a few minor comments - feel free to merge after looking at those

File: src/collator/collator.py (hunk 3 above)
Comment (sabraaap) [reply src/collator/collator.py:75]: Yep, only need the update epoch info for curriculum collator actually (so not the single task one and not the round robin one), just passing when the method is called on the other collators seemed like the easiest thing to do (not very elegant oc, so if you want to do something with it go ahead!)

File: scripts/train_agent_babyai.py (hunk 1 above)
Comment (sabraaap) [reply scripts/train_agent_babyai.py:59]: Let me know if that's not enough and you want it somewhere else too

**Review by maxtaylordavies — APPROVED** [review]
